lib/kafo/param.rb

```diff
@@ -34,8 +35,34 @@ def unset_value
       @value     = nil
     end
 
+    # For literal default values, only use 'manifest_default'. For variable values, use the value
+    # loaded back from the dump in 'default'.
+    def default
+      @type.typecast(dump_default_needed? ? @default : manifest_default)
+    end
+
+    def default=(default)
+      default = nil if ['UNSET', :undef].include?(default)
+      @default = default
+    end
+
+    # manifest_default may be a variable ($foo::params::bar) and need dumping from Puppet to get
+    # the actual default value
+    def dump_default_needed?
+      manifest_default.to_s.start_with?('$')
+    end
+
     def dump_default
-      @type.dump_default(default)
+      @type.dump_default(manifest_default_params_variable)
+    end
+
+    def manifest_default=(default)
+      default = nil if ['UNSET', :undef].include?(default)
+      @manifest_default = default
+    end
+
+    def manifest_default_params_variable
+      manifest_default[1..-1] if manifest_default.to_s.start_with?('$')
```

Comment:
`if dump_default_needed?` or rename/alias the method so it makes sense in this context